### PR TITLE
userprofile: initial commit

### PR DESCRIPTION
**Pull request by SamiHiltunen**
* Initial commit for Invenio-UserProfile.

Signed-off-by: Sami Hiltunen <sami.mikael.hiltunen@cern.ch>

Still work in progress

### CODE REVIEW

File: AUTHORS.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    waive the privileges and immunities granted to it by virtue of its status
+    as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+
+Authors
+=======
+
+Invenio module that adds userprofiles to the platform.
+
+- CERN <info@invenio-software.org>
```
Comment (lnielsen) [inline AUTHORS.rst]: Just remove this line

File: invenio_userprofile/version.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+"""Version information for Invenio-UserProfile.
+
+This file is imported by ``invenio_userprofile.__init__``,
+and parsed by ``setup.py``.
+"""
+
+from __future__ import absolute_import, print_function
+
+__version__ = "0.1.0.dev20150000"
```
Comment (lnielsen) [inline invenio_userprofile/version.py]: Let's start out with ``1.0.0.dev20150000``

**Comment by SamiHiltunen** [comment]
Fixed the points @lnielsen brought up. This depends on inveniosoftware/flask-menu#39 being released.

File: MANIFEST.in [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+
+# TODO: Generate this manifest file by running the following commands:
+#
+#  git init
+#  git add -A
+#  pip install -e .[all]
+#  check-manifest -u
```
Comment (jirikuncar) [inline MANIFEST.in]: @SamiHiltunen please run these commands and then remove the TODO.

File: examples/app.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+u"""Minimal Flask application example for development.
+
+Install the Invenio default theme
+
+You should execute these commands in the examples-directory.
+
+.. code-block:: console
+
+   $ pip install -e \
```
Comment (jirikuncar) [inline examples/app.py]: maybe you can add it to `requirements.txt`

File: examples/app.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    InvenioTheme(app)
+InvenioAccounts(app)
+InvenioUserProfile(app)
+
+
+@app.route('/')
+def index():
+    return redirect(url_for('invenio_userprofile.profile'))
+
+if __name__ == "__main__":
```
Comment (jirikuncar) [inline examples/app.py]: you can remove this as we are using `flask` to run `app.py`.

File: invenio_userprofile/api.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+def _get_current_userprofile():
+    if current_user.is_anonymous():
+        return AnonymousUserProfile()
+
+    profile = UserProfile.query.get(current_user.id)
+    if not profile:
+        with db.session.begin_nested():
+            profile = UserProfile()
+            profile.user_id = current_user.id
```
Comment (jirikuncar) [inline invenio_userprofile/api.py]: shall we use `get_id()` on `current_user` (seehttps://flask-login.readthedocs.org/en/latest/#your-user-class)?

```python
profile = UserProfile(user_id=current_user.get_id())
```

File: invenio_userprofile/ext.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+from __future__ import absolute_import, print_function
+
+from .api import current_userprofile
+
+from .views import blueprint
+
+
+class InvenioUserProfile(object):
+
```
Comment (jirikuncar) [inline invenio_userprofile/ext.py]: - remove line before class docstring
- please get latest `pep257` package to verify

File: invenio_userprofile/templates/invenio_userprofile/index.html [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{%- extends config.USERPROFILE_BASE_TEMPLATE %}
```
Comment (jirikuncar) [inline invenio_userprofile/templates/invenio_userprofile/index.html]: missing licence header

File: invenio_userprofile/templates/invenio_userprofile/index.html [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+
+{%- block page_body %}
+{{_('Welcome to %(module_name)s', module_name=module_name)}}
+{%- endblock %}
+
```
Comment (jirikuncar) [inline invenio_userprofile/templates/invenio_userprofile/index.html]: empty line can be removed

File: invenio_userprofile/templates/invenio_userprofile/settings/profile.html [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    <label class="control-label" for="first_name">{{ text_first_name }}</label>
+    {{ render_field(form.first_name, autofocus=True, placeholder=text_first_name) }}
+</div>
+<div class="form-group">
+    <label class="control-label" for="last_name">{{ text_last_name }}</label>
+    {{ render_field(form.last_name, autofocus=True, placeholder=text_last_name) }}
+</div>
+
+<div class="form-actions">
+    <button type="submit" class="btn btn-primary">{{ text_update_profile}}</button>
```
Comment (jirikuncar) [inline invenio_userprofile/templates/invenio_userprofile/settings/profile.html]: `{{ text_update_profile }}`

File: misc/header.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
```
Comment (jirikuncar) [inline misc/header.py]: `misc` folder can be removed

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    keywords='invenio profile account user',
+    license='GPLv2',
+    author='CERN',
+    author_email='info@invenio-software.org',
+    url='https://github.com/inveniosoftware/invenio-userprofile',
+    packages=packages,
+    zip_safe=False,
+    include_package_data=True,
+    platforms='any',
+    entry_points={},
```
Comment (jirikuncar) [inline setup.py]: it should include translations and models

File: tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    """Flask application fixture."""
+    app = Flask('testapp')
+    app.config.update(
+        TESTING=True,
+        LOGIN_DISABLED=False,
+        SECRET_KEY='testing_key',
+        TEST_USER_EMAIL='test_user@example.com',
+        TEST_USER_PASSWORD='test_password',
+        WTF_CSRF_ENABLED=False,
+        SQLALCHEMY_DATABASE_URI='sqlite://',  # in memory db
```
Comment (jirikuncar) [inline tests/conftest.py]: we should probably check `os.environ` for `SQLALCHEMY_DATABASE_URI` so we can test all databases on Travis CI.

File: tests/test_models.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+from test_validators import test_usernames
+
+
+def test_userprofile(app):
+    """Test UserProfile model."""
+    profile = UserProfile()
+
+    # Check the username validator works on the model
+    profile.username = test_usernames['valid']
+    try:
```
Comment (jirikuncar) [inline tests/test_models.py]: You probably want to test that the exception is always raised.

```python
with pytest.raises(ValueError):
    profile.username = test_usernames['invalid_characters']
```

File: tests/test_validators.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+
+def test_validate_username(app):
+    """Test username validator."""
+    # Goodname can contain letters, numbers and starts with a letter
+    validate_username(test_usernames['valid'])
+
+    # Can't start with a number
+    try:
```
Comment (jirikuncar) [inline tests/test_validators.py]: see https://github.com/inveniosoftware/invenio-userprofile/pull/1/files#r43604796

File: tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+            # Whitespace should be trimmed
+            client.post(url_for('invenio_userprofile.profile'), data=dict(
+                username='{0} '.format(test_usernames['valid']),
+                first_name='Valid  ',
+                last_name='Name  '))
+            resp = client.get(url_for('invenio_userprofile.profile'))
+            assert resp.status_code == 200
+            assert test_usernames['valid'] in resp.data
+            assert 'Valid' in resp.data
```
Comment (jirikuncar) [inline tests/test_views.py]: maybe you should better check that `'Valid  ' not in resp.data`.

File: invenio_userprofile/models.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+    id = db.Column(db.Integer, primary_key=True)
+
+    user_id = db.Column(db.Integer, db.ForeignKey(User.id))
+
+    username = db.Column(db.String(255), unique=True)
+
+    first_name = db.Column(db.String(255))
+
+    last_name = db.Column(db.String(255))
```
Comment (kaplun) [inline invenio_userprofile/models.py]: Hi! `first_name`, `last_name` is minefield in our international environment. What about simply: `preferred_name` or `fullname`?

Comment (kaplun) [inline invenio_userprofile/models.py]: See also: http://ux.stackexchange.com/a/15778

Comment (lnielsen) [inline invenio_userprofile/models.py]: +1 for only ``fullname``. Will that be fine for Inspire? I remember you had problems with the ORCID login.

Comment (kaplun) [inline invenio_userprofile/models.py]: In the case of INSPIRE we can simply copy whatever info comes from ORCID indeed. So there we could concatenate on the fly its field and make our fullname, I guess. @jmartinm?

Comment (SamiHiltunen) [inline invenio_userprofile/models.py]: Very good points. I removed first_name and last_name and replaced them with full_name.

File: tests/test_views.py (hunk 16 above)
Comment (SamiHiltunen) [inline tests/test_views.py]: Definitely makes it more clear :+1:

**Comment by SamiHiltunen** [comment]
I believe all the points brought up have been addressed now.

File: RELEASE-NOTES.rst [hunk 18]
```diff
@@ -0,0 +1,36 @@
+============================
```
Comment (jirikuncar) [inline RELEASE-NOTES.rst]: `=`

File: RELEASE-NOTES.rst [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+ Invenio-UserProfiles v0.1.0
+============================
```
Comment (jirikuncar) [inline RELEASE-NOTES.rst]: add `=`

File: requirements.txt [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+-e git+https://github.com/inveniosoftware/invenio-assets.git#egg=invenio-assets
```
Comment (jirikuncar) [inline requirements.txt]: it should be probably moved to `examples` directory

**Comment by SamiHiltunen** [comment]
@jirikuncar Please enable Travis for the repository.

**Comment by jirikuncar** [comment]
@SamiHiltunen done. You can re-push.

File: setup.py [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+    'Babel>=1.3',
+]
+
+install_requires = [
+    'Flask-BabelEx>=0.9.2',
+    'Flask-Mail>=0.9.1',
+    'Flask-Menu>=0.4.0',
+    'Flask-WTF>=0.10.2',
+    'invenio-accounts>=0.2.0',
+    'invenio-db>=1.0.0a4',
```
Comment (jirikuncar) [inline setup.py]: please have a look at `invenio:setup.py` to install `invenio-db` package from extras

**Comment by jirikuncar** [comment]
- [ ] fix failing Travis CI
- [ ] install invenio-db[...] from extras

**Pull request merged**